public/app/plugins/datasource/tempo/metricsSummary.ts

```diff
@@ -0,0 +1,294 @@
+  errorSpanCount?: string;
+  p50: string;
+  p90: string;
+  p95: string;
+  p99: string;
+  series: Series[];
+};
+
+type Series = {
+  key: string;
+  value: {
+    type: number;
+    n?: number;
+    f?: number;
+    s?: string;
+    b?: string;
+    d?: string;
+    status?: number;
+    kind?: number;
+  };
+};
+
+type MetricsData = {
+  spanCount: string;
+  errorPercentage: string;
+  p50: string;
+  p90: string;
+  p95: string;
+  p99: string;
+  [key: string]: string | number;
+};
+
+export function createTableFrameFromMetricsSummaryQuery(
+  data: MetricsSummary[],
+  targetQuery: string,
+  instanceSettings: DataSourceInstanceSettings,
+  state: LoadingState
+) {
+  let frame;
+
+  if (state === LoadingState.Error) {
+    frame = emptyResponse;
+  } else if (state === LoadingState.Loading) {
+    frame = new MutableDataFrame({
+      name: 'Metrics Summary',
+      refId: 'metrics-summary',
+      fields: [
+        { name: 'state', type: FieldType.string, config: { displayNameFromDS: 'State', custom: { width: 300 } } },
+      ],
+      meta: {
+        preferredVisualisationType: 'table',
+      },
+    });
+    frame.add({
+      state: 'Loading...',
+    });
+  } else if (state === LoadingState.Done && data?.length) {
+    const dynamicMetrics: Record<string, FieldDTO> = {};
+    data.map((res: MetricsSummary) => {
+      const configQuery = getConfigQuery(res.series, targetQuery);
+      res.series.map((series: Series) => {
+        return !dynamicMetrics[series.key]
+          ? (dynamicMetrics[series.key] = {
+              name: `${series.key}`,
+              type: FieldType.string,
+              config: getConfig(series, configQuery, instanceSettings),
+            })
+          : dynamicMetrics[series.key];
+      });
+    });
+
+    frame = new MutableDataFrame({
+      name: 'Metrics Summary',
+      refId: 'metrics-summary',
+      fields: [
+        ...Object.values(dynamicMetrics).sort((a, b) => a.name.localeCompare(b.name)),
+        {
+          name: 'kind',
+          type: FieldType.string,
+          config: { displayNameFromDS: 'Kind', custom: { width: 150 } },
+        },
+        {
+          name: 'spanCount',
+          type: FieldType.string,
+          config: { displayNameFromDS: 'Span count', custom: { width: 150 } },
+        },
+        {
+          name: 'errorPercentage',
+          type: FieldType.string,
+          config: { displayNameFromDS: 'Error', unit: 'percent', custom: { width: 150 } },
+        },
+        getPercentileRow('p50'),
+        getPercentileRow('p90'),
+        getPercentileRow('p95'),
+        getPercentileRow('p99'),
+      ],
+      meta: {
+        preferredVisualisationType: 'table',
+      },
+    });
+
+    if (!data?.length) {
+      return emptyResponse;
+    }
+
+    const metricsData = data.map(transformToMetricsData);
+    for (const trace of metricsData) {
+      frame.add(trace);
+    }
+    frame = sortDataFrame(frame, 0);
+  }
+
+  if (!frame) {
```

Comment:
I guess if we move `if (!data?.length) {` to the top we can then be sure that `frame` always exists if there is `data`?